cryodrgn/source.py

```diff
@@ -0,0 +1,400 @@
+        ]
+        n_workers = min(n_workers, len(self.source_intervals))
+
+        super().__init__(L=self.ny, n=self.nz, n_workers=n_workers, *args, **kwargs)
+
+    def _images(self, indices: np.ndarray):
+        def load_single_mrcs(
+            data: np.ndarray,
+            src: MRCFileSource,
+            src_indices: np.ndarray,
+            tgt_indices: np.ndarray,
+        ):
+            src._images(indices=src_indices, data=data, tgt_indices=tgt_indices)
+
+        data = np.empty((len(indices), self.L, self.L), dtype=self.dtype)
+
+        with futures.ThreadPoolExecutor(self.n_workers) as executor:
+            to_do = []
+            for source_i, (source_start_index, source_end_index) in enumerate(
+                self.source_intervals
+            ):
+                tgt_indices = np.where(
+                    (source_start_index <= indices) & (indices < source_end_index)
+                )[0]
+                src_indices = indices[tgt_indices] - source_start_index
+                src_images = src_indices.size
+
+                if src_images > 0:
+                    future = executor.submit(
+                        load_single_mrcs,
+                        data=data,
+                        src=self.sources[source_i],
+                        src_indices=src_indices,
+                        tgt_indices=tgt_indices,
+                    )
+                    to_do.append(future)
+
+            for future in futures.as_completed(to_do):
+                res, exc = future.result(), future.exception()
+                if exc is not None:
+                    raise exc
+
+        return data
+
+
+class _MRCDataFrameSource(ImageSource):
+    def __init__(
+        self, df: pd.DataFrame, datadir: str = "", n_workers: int = 1, *args, **kwargs
+    ):
+        assert "__mrc_index" in df.columns
+        assert "__mrc_filename" in df.columns
+        self.df = df
+
+        self.df["__mrc_filepath"] = self.df["__mrc_filename"].apply(
+            lambda filename: os.path.join(datadir, filename)
+        )
+
+        # Peek into the first mrc file to get image size
+        L = MRCFileSource(self.df["__mrc_filepath"][0]).L
+        super().__init__(
+            L=L,
+            n=len(self.df),
+            filenames=df["__mrc_filename"],
+            n_workers=n_workers,
+            *args,
+            **kwargs,
+        )
+
+    def _images(self, indices: np.ndarray):
+        def load_single_mrcs(filepath, df):
+            src = MRCFileSource(filepath)
+            # df.index indicates the positions where the data needs to be inserted -> return for use by caller
+            return df.index, src._images(df["__mrc_index"])
+
+        data = np.empty((len(indices), self.L, self.L), dtype=self.dtype)
+
+        # Create a DataFrame corresponding to the indices we're interested in
+        df = self.df.iloc[indices].reset_index(drop=True)
+        groups = df.groupby("__mrc_filepath")
```

Comment:
This will be too expensive if you have 100M images and you're fetching O(1) images for a batch.

If you're going to do this you'll need to precompute it when constructing this object.

In general, I think you need to be testing/profiling on a real-world big dataset (that @zhonge  can point you to), not just on toy examples with a few hundred images.

Reply:
I take this back; I missed that you're first creating a dataframe for just this batch; I think this is fine (and nice).

Maybe rename `df` to `batch_df`.

Reply:
Renamed in b2f85194f01817c24ca579a66f6014628930d29e